libavfilter/sf_text2graphicsub.c

```diff
@@ -0,0 +1,441 @@
+            ass_add_font(context->library, tag->value,
+                 (char *)st->codecpar->extradata,
+                 st->codecpar->extradata_size);
+        }
+        avformat_close_input(&video);
+    }
+
+    context->renderer = ass_renderer_init(context->library);
+    ass_set_pixel_aspect(context->renderer, 1);
+    ass_set_shaper(context->renderer, 0);
+    ass_set_fonts(context->renderer, NULL, NULL, 1, NULL, 1);
+
+    init_palettizecontext(&context->palettize_context);
+
+    ff_draw_init(&context->draw_context, AV_PIX_FMT_BGRA, FF_DRAW_PROCESS_ALPHA);
+
+    return 0;
+}
+
+static int config_input(AVFilterLink *inlink)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    ASS_Context *context = ctx->priv;
+    if (!context->size.width)  context->size.width  = inlink->w;
+    if (!context->size.height) context->size.height = inlink->h;
+    if (!context->size.height || !context->size.width) {
+        av_log(NULL, AV_LOG_ERROR, "A positive height and width are required to render subtitles\n");
+        return AVERROR_EXIT;
+    }
+    ass_set_frame_size(context->renderer, context->size.width, context->size.height);
+    ass_set_storage_size(context->renderer, context->size.width, context->size.height);
+
+    return 0;
+}
+
+static int config_output(AVFilterLink *outlink)
+{
+    AVFilterContext *ctx = outlink->src;
+    ASS_Context *context = ctx->priv;
+
+    outlink->w = context->size.width;
+    outlink->h = context->size.height;
+
+    return 0;
+}
+
+static int filter_frame(AVFilterLink *inlink, AVFrame *frame)
+{
+    const AVFilterContext *ctx = inlink->dst;
+    ASS_Context *context = ctx->priv;
+    AVFilterLink *outlink = ctx->outputs[0];
+    const int64_t start_time = av_rescale_q(frame->subtitle_timing.start_pts, AV_TIME_BASE_Q, av_make_q(1, 1000));
+    const int64_t duration   = av_rescale_q(frame->subtitle_timing.duration, AV_TIME_BASE_Q, av_make_q(1, 1000));
+    
+    if (!context->track) {      // Init track
+        if (frame->subtitle_header)
+            context->track = ass_read_memory(context->library,
+                (char *) frame->subtitle_header->data,
+                frame->subtitle_header->size, NULL);
+        else context->track = ass_new_track(context->library);
+    }
+
+    for (int r = 0; r < frame->num_subtitle_areas; r++)
+    {
+        ASS_Image *image;
+        AVSubtitleArea *area = frame->subtitle_areas[r];
+        if (area->type == SUBTITLE_BITMAP) continue;
+        if (area->ass == NULL) continue;
+        
+        ass_process_chunk(context->track, area->ass, strlen(area->ass), start_time, duration);
+        image = ass_render_frame(context->renderer, context->track, start_time + duration / 2, NULL);
+        if (image == NULL) {
+            av_log(NULL, AV_LOG_WARNING, "failed to render ass: %s\n", area->ass);
+            return 0;
+        }
+
+        ass_image_to_area_palletization(context, image, area);
+        area->type = SUBTITLE_BITMAP;
```

Comment:
and this must go to the top of the loop. We need to make sure that all areas are bitmap type

Reply:
oh - no, we must not emit any empty areas either.